.github/workflows/deploy-universal.yml

```diff
@@ -0,0 +1,102 @@
+name: Deployment (Universal)
+
+on:
+  workflow_dispatch:
+    inputs:
+      environment:
+        description: 'Environment to deploy to'
+        required: true
+        default: 'stage'
+        type: choice
+        options: ['stage', 'prod']
+      delete-volumes:
+        description: 'Delete volumes?'
+        type: boolean
+        default: false
+      restore-db:
+        description: 'Restore DB?'
+        type: boolean
+        default: false
+      flush-redis:
+        description: 'Clear redis?'
+        type: boolean
+        default: false
+      build-flag:
+        type: choice
+        description: docker-compose additional build flag
+        default: "None"
+        options:
+          - "None"
+          - "--no-cache"
+
+jobs:
+  deployment:
+    concurrency: ${{ github.event.inputs.environment }}
+    runs-on: [self-hosted, "${{ github.event.inputs.environment }}"]
+    environment: ${{ github.event.inputs.environment == 'stage' && 'Staging' || 'Production' }}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Verify environment
+        run: |
+          (diff -wB <(grep -vE '^\s*#' /opt/deployment-specific-files/.env | cut -d '=' -f 1 | sort) \
+          <(grep -vE '^\s*#' .env.sample | cut -d '=' -f 1 | sort)) ||\
+          (echo "Environment variables mismatch" &&\
+          echo "Update /opt/deployment-specific-files/.env on the host or .env.sample in your branch" &&\
+          exit 1)
+
+      - name: Copy configuration
+        run: |
+          cp -f /opt/deployment-specific-files/chain.pem ./caddy/certs
+          cp -f /opt/deployment-specific-files/priv.pem ./caddy/certs
+          cp -f /opt/deployment-specific-files/.env ./
+
+      - name: Build
+        env:
+          # replace None with an empty string
+          BUILD_FLAG: ${{ github.event.inputs.build-flag != 'None' && github.event.inputs.build-flag || '' }}
+        run: docker-compose -f docker-compose.ssl.local.yml build $BUILD_FLAG
+
+      - name: Download DB dump
+        id: get-db
+        if: ${{ github.event.inputs.restore-db }}
+        uses: ./.github/actions/get-db
+        with:
+          workflow: ${{ github.event.inputs.environment == 'stage' && 'restore-sanitized-db.yml' || 'backup-db.yml' }}
+          zip-password: ${{ github.event.inputs.environment == 'stage' && secrets.STAGING_BACKUP_PASSWORD || secrets.BACKUP_PASSWORD }}
+
+      - name: Copy everything to /orbitar
+        run: |
+          mkdir -p /orbitar
+          rm -rf /orbitar/*
+          cp -r ./* /orbitar/
+          cd /orbitar
+
+      - name: Delete volumes
+        if : ${{ github.event.inputs.delete-volumes }}
+        run: docker-compose -f docker-compose.ssl.local.yml down -v
+
+      - name: Restore DB
+        if : ${{ github.event.inputs.restore-db }}
+        run: |
+          docker-compose -f docker-compose.ssl.local.yml stop backend
```

Comment:
This one recreates all dependent containers. Thus, if they were deleted in the previous step, they may now exist prematurely.

Reply:
I don't think it **re**creates them, just creates if they were previously deleted. And they will be used anyway in the latter step.

Reply:
Recreates but not restarts. If we were to do some stuff to the volumes before explicitly recreating them, this could be shooting yourself in the foot eventually. 
Not insisting on changing it though

Reply:
I'm not entirely sure what's the concern here is, but I'm open for suggestion.
My goal here was to ensure that backend is stopped before manipulating redis or mysql.

Reply:
I'm not sure it's going to work.

```
docker-compose -f docker-compose.ssl.local.yml stop backend
docker-compose -f docker-compose.ssl.local.yml up -d mysql --wait
```
The second command will recreate the volumes anyway.

Reply:
Just tried it locally:
```
docker-compose -f docker-compose.ssl.local.yml up -d mysql --wait
[+] Running 6/7
 ⠿ Network orbitar_default          Created                                                                                                                                                           0.0s
 ⠿ Volume "orbitar_caddy_data"      Created                                                                                                                                                           0.0s
 ⠿ Volume "orbitar_caddy_config"    Created                                                                                                                                                           0.0s
 ⠿ Volume "orbitar_mysql_db"        Created                                                                                                                                                           0.0s
 ⠿ Volume "orbitar_redis_data"      Created                                                                                                                                                           0.0s
 ⠿ Volume "orbitar_frontend_build"  Created                                                                                                                                                           0.0s
 ⠿ Container orbitar-mysql-1        Waiting                                                                                                                                                          13.6s
```

Reply:
Yup, looks like it, sorry.